apollo-router/src/plugins/external.rs

```diff
@@ -88,196 +254,194 @@ struct Conf {
-                        // updated by our co-processor and decide if we should proceed or stop.
-
-                        if matches!(co_processor_output.control, Control::Break(_)) {
-                            // Ensure the code is a valid http status code
-                            let code = co_processor_output.control.get_http_status()?;
-
-                            let res = if !code.is_success() {
-                                router::Response::error_builder()
-                                    .errors(vec![Error {
-                                        message: co_processor_output
-                                            .body
-                                            .unwrap_or(serde_json::Value::Null)
-                                            .to_string(),
-                                        ..Default::default()
-                                    }])
-                                    .status_code(code)
-                                    .context(request.context)
-                                    .build()?
-                            } else {
-                                router::Response::builder()
-                                    .data(
-                                        co_processor_output
-                                            .body
-                                            .unwrap_or(serde_json::Value::Null)
-                                            .to_string(),
-                                    )
-                                    .status_code(code)
-                                    .context(request.context)
-                                    .build()?
-                            };
-                            return Ok(ControlFlow::Break(res));
-                        }
-
-                        // Finally, process our reply and act on the contents. Our processing logic is
-                        // that we replace "bits" of our incoming request with the updated bits if they
-                        // are present in our co_processor_output.
-
-                        let new_body = match co_processor_output.body {
-                            Some(bytes) => Body::from(serde_json::to_vec(&bytes)?),
-                            None => Body::from(b_bytes),
+                    tracing::debug!(?payload, "externalized output");
+                    request.context.enter_active_request().await;
+                    let co_processor_result = payload.call(http_client, &coprocessor_url).await;
+                    request.context.leave_active_request().await;
+                    tracing::debug!(?co_processor_result, "co-processor returned");
+                    let co_processor_output = co_processor_result?;
+
+                    validate_coprocessor_output(&co_processor_output, PipelineStep::RouterRequest)?;
+                    // unwrap is safe here because validate_coprocessor_output made sure control is available
```

Comment:
Maybe return a reference to control in `validate_coprocessor_output rather than unwrap.

Reply:
Controls wont be available in pipelinestep::XXXResponse, so `validate_coprocessor_output` would return an Option<Control> anyway :/

Reply:
sure!

Reply:
all good